source/common/upstream/load_balancer_impl.cc

```diff
@@ -14,18 +14,50 @@ const std::vector<HostPtr>& LoadBalancerBase::hostsToUse() {
     return host_set_.hosts();
   }
 
-  uint64_t threshold =
+  uint64_t global_panic_threshold =
       std::min(100UL, runtime_.snapshot().getInteger("upstream.healthy_panic_threshold", 50));
   double healthy_percent =
       (host_set_.healthyHosts().size() / static_cast<double>(host_set_.hosts().size())) * 100;
 
   // If the % of healthy hosts in the cluster is less than our panic threshold, we use all hosts.
-  if (healthy_percent < threshold) {
+  if (healthy_percent < global_panic_threshold) {
     stats_.upstream_rq_lb_healthy_panic_.inc();
     return host_set_.hosts();
-  } else {
-    return host_set_.healthyHosts();
   }
+
+  // Attempt to do zone aware routing if there are at least 2 upstream zones and it's enabled.
+  if (stats_.upstream_zone_count_.value() > 1 &&
```

Comment:
I would also short circuit id the # of local zone hosts is zero. This happens for all non-SDS cluster types, and if SDS is misconfigured.